### DISCUSSION TITLE
vipssave problem with a large image

### DISCUSSION THREAD

**Discussion by pdbourke** [libvips/libvips#3934] [General]
This is related, I suspect, to the last item in the other discussion titled "resize 16bit tiff images", but I've investigated some more.

I have two images: scroll3_8.tif and scroll3_8_10ppm.tif
The first file was created using vips from a 16bit version using colourspace rgb16 to srgb conversion.

> vipsheader scroll3_8.tif
scroll3_8.tif: 1135360x440000 uchar, 3 bands, srgb, tiffload

> vipsheader scroll3_8_10ppm.tif
scroll3_8_10ppm.tif: 283840x110000 uchar, 3 bands, srgb, tiffload

The later is 1/4 the size of the first on each axis.
It was created as follows:

> vips resize scroll3_8.tif scroll3_8_10ppm.tif 0.25 --vips-progress
vips.exe temp-1: 283840 x 110000 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
vips.exe temp-1: done in 8.83e+03s

Converting the second file to vips format works, takes about 12 minutes.

> vips vipssave scroll3_8_10ppm.tif scroll3_8_10ppm.v --vips-progress
vips.exe scroll3_8_10ppm.v: 283840 x 110000 pixels, 32 threads, 283840 x 1 tiles, 640 lines in buffer
vips.exe scroll3_8_10ppm.v: done in 731s

However trying to save the first image to vips format and it hasn't moved from 0% in 8 hours.

> vips vipssave .\scroll3_8.tif scroll3_8.v --vips-progress
vips.exe scroll3_8.v: 1135360 x 440000 pixels, 32 threads, 1135360 x 1 tiles, 640 lines in buffer
vips.exe scroll3_8.v: 0% complete

Ideas?

**Comment by jcupitt** [comment]
You don't need `vipssave` (I suspect it doesn't work well, it's only there for completeness with some language bindings), just do:

```
vips copy .\scroll3_8.tif scroll3_8.v --vips-progress
```

**Comment by pdbourke** [comment]
> You don't need vipssave (I suspect it doesn't work well, it's only there for completeness with some language bindings), just do:
> vips copy .\scroll3_8.tif scroll3_8.v --vips-progress

Same result with copy and vipssave.

This is getting bizarre (or interesting depend on how you look at it),
if I do this
    vips resize scroll3_8.tif N.v N --vips-progress
for various value of N, I get the following

N       outcome
0.5 success
0.75 success
0.9 success
0.98 success
0.985 success (1118330 x 433400)
0.986 never completes (1119465 x 433840)
0.987 never completes
0.989 never completes
0.99 never completes

Some magic threshold is being hit??

> vipsheader -a scroll3_8.tif
scroll3_8.tif: 1135360x440000 uchar, 3 bands, srgb, tiffload

The reason why I want to create a .v file is that it is easy for me to
read in order to perform some of the custom processing the project
requires.

--
Paul Bourke
paulbourke.net
***@***.***
+61 433338325)

**Comment by pdbourke** [comment]
And just to verify it's some as yet unknown limit, I tried the same
experiment with a larger image and the cut-off point was between a 79%
and 80% reduction.

On Mon, 15 Apr 2024 at 09:49, Paul Bourke ***@***.***> wrote:
>
> > You don't need vipssave (I suspect it doesn't work well, it's only there for completeness with some language bindings), just do:
> > vips copy .\scroll3_8.tif scroll3_8.v --vips-progress
>
> Same result with copy and vipssave.
>
> This is getting bizarre (or interesting depend on how you look at it),
> if I do this
>     vips resize scroll3_8.tif N.v N --vips-progress
> for various value of N, I get the following
>
> N       outcome
> 0.5 success
> 0.75 success
> 0.9 success
> 0.98 success
> 0.985 success (1118330 x 433400)
> 0.986 never completes (1119465 x 433840)
> 0.987 never completes
> 0.989 never completes
> 0.99 never completes
>
> Some magic threshold is being hit??
>
> > vipsheader -a scroll3_8.tif
> scroll3_8.tif: 1135360x440000 uchar, 3 bands, srgb, tiffload
>
> The reason why I want to create a .v file is that it is easy for me to
> read in order to perform some of the custom processing the project
> requires.
>
> --
> Paul Bourke
> paulbourke.net
> ***@***.***
> +61 433338325)



-- 
Paul Bourke
paulbourke.net
***@***.***
+61 433338325

**Reply by jcupitt** [reply]
I tried here on ubuntu 23.10 and libvips 8.15:

```
$ vips replicate ../k2.jpg x.tif[bigtiff] 1000 10 
$ vipsheader x.tif
x.tif: 1450000x20480 uchar, 3 bands, srgb, tiffload
$ ls -l x.tif
-rw-r--r-- 1 john john 89088002252 Apr 15 11:25 x.tif
$ vips resize x.tif x.v 0.99 --vips-progress
vips x.v: 1435500 x 20275 pixels, 32 threads, 1435500 x 16 tiles, 640 lines in buffer
vips x.v: 3% complete
```

So a 90gb image with about the same width. libvips works in chunks down from the top, so I'm hoping there's no need to make one the full height, you just have to make it very wide. It seems to work (though it's not quick!).

I think we need to try:

- another platform (ie. is this an issue with the windows build, or with libvips generally)
- you could try disabling vectorisation using the `--vips-novector` flag
- could you try those exact commands above (though with a different test jpg)

**Comment by pdbourke** [comment]
To start with, thanks for your help and advice.

1. Late here now, tomorrow I'll see if the height does matter,
replicate your 1450000x20480.

2. I've requested an update to the latest version of vips on Windows.
I'll then test that, should be in the next day or two.

3. Then I'll try WSL, never done that but we'll see how it goes.

4. I've asked them to send me a drive with the three image files.
Gives me an excuse to buy a well endowed Macstudio. :-) But I won't be
able to test that until May 6 when I come back from 2 weeks in New
Zealand.

On Mon, 15 Apr 2024 at 20:32, John Cupitt ***@***.***> wrote:
>
> I tried here on ubuntu 23.10 and libvips 8.15:
>
> $ vips replicate ../k2.jpg x.tif[bigtiff] 1000 10
> $ vipsheader x.tif
> x.tif: 1450000x20480 uchar, 3 bands, srgb, tiffload
> $ ls -l x.tif
> -rw-r--r-- 1 john john 89088002252 Apr 15 11:25 x.tif
> $ vips resize x.tif x.v 0.99 --vips-progress
> vips x.v: 1435500 x 20275 pixels, 32 threads, 1435500 x 16 tiles, 640 lines in buffer
> vips x.v: 3% complete
>
> So a 90gb image with about the same width. libvips works in chunks down from the top, so I'm hoping there's no need to make one the full height, you just have to make it very wide. It seems to work (though it's not quick!).
>
> I think we need to try:
>
> another platform (ie. is this an issue with the windows build, or with libvips generally)
> you could try disabling vectorisation using the --vips-novector flag
> could you try those exact commands above (though with a different test jpg)
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***>



-- 
Paul Bourke
paulbourke.net
***@***.***
+61 433338325

**Reply by jcupitt** [reply]
I've not been able to reproduce this issue here, so let's talk again in May.

WSL is really easy and works surprisingly well, I'm sure it'll be fine.
